### PR TITLE
Sdk 5 init

### PR DESCRIPTION
**Pull request by matemax**

### CODE REVIEW

File: lunavl/sdk/estimators/face_estimators/orientation_mode.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        """
+        Init.
+
+        Args:
+            coreOrientationModeEstimator: core orientation mode estimator
+        """
+        super().__init__(coreOrientationModeEstimator)
+
+    @CoreExceptionWrap(LunaVLError.EstimationOrientationModeError)
+    def estimate(self, coreImage: CoreImage) -> OrientationMode:
```
Comment (matemax) [inline lunavl/sdk/estimators/face_estimators/orientation_mode.py]: We don't use  CoreImage  as input user data

File: lunavl/sdk/estimators/face_estimators/orientation_mode.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+from lunavl.sdk.base import BaseEstimation
+from ..base import BaseEstimator
+
+
+class OrientationType(Enum):
+    """
+    Enum for orientation type
+    """
+
+    LEFT = "Left"
```
Comment (matemax) [inline lunavl/sdk/estimators/face_estimators/orientation_mode.py]: description please

File: lunavl/sdk/estimators/face_estimators/orientation_mode.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
+"""
```
Comment (matemax) [inline lunavl/sdk/estimators/face_estimators/orientation_mode.py]: it is not face estimator

File: lunavl/sdk/estimators/face_estimators/orientation_mode.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+            orientation type
+        """
+        orientationType = cls(coreOrientationMode.name)
+        return orientationType
+
+    def __repr__(self):
+        return self.value
+
+
+class OrientationMode(BaseEstimation):
```
Comment (matemax) [inline lunavl/sdk/estimators/face_estimators/orientation_mode.py]: I think that class is useless and estimator  must returns `OrientationType`

File: lunavl/sdk/faceengine/setting_provider.py [hunk 5]
```diff
@@ -240,7 +240,7 @@ def getValueAsInt(self, name: str) -> int:
         Returns:
             a int
         """
-        return self.getValue(name).asInt()
+        return self.getValue(name)[0]
```
Comment (matemax) [inline lunavl/sdk/faceengine/setting_provider.py]: check other functions

File: tests/test_estimate_descriptor.py [hunk 6]
```diff
@@ -27,7 +27,7 @@
 from tests.detect_test_class import VLIMAGE_SMALL
 from tests.resources import WARP_WHITE_MAN, HUMAN_WARP, WARP_CLEAN_FACE, BAD_THRESHOLD_WARP
 
-EFDVa = EXISTENT_FACE_DESCRIPTOR_VERSION_ABUNDANCE = [46, 52, 54, 56]
+EFDVa = EXISTENT_FACE_DESCRIPTOR_VERSION_ABUNDANCE = [54, 56]
```
Comment (matemax) [inline tests/test_estimate_descriptor.py]: 57?

File: lunavl/sdk/estimator_collections.py [hunk 7]
```diff
@@ -44,6 +46,10 @@ class FaceEstimator(Enum):
     Mask = 10
     #: glasses estimator
     Glasses = 11
+    #: liveness v1 estimator
+    LivebessV1 = 12
```
Comment (DeusAnimo) [inline lunavl/sdk/estimator_collections.py]:
```suggestion
    LivenessV1 = 12
```

File: lunavl/sdk/estimator_collections.py [hunk 8]
```diff
@@ -183,6 +195,10 @@ def initEstimator(self, estimator: FaceEstimator) -> None:
             self._maskEstimator = self._faceEngine.createMaskEstimator()
         elif estimator == FaceEstimator.Glasses:
             self._glassesEstimator = self._faceEngine.createGlassesEstimator()
+        elif estimator == FaceEstimator.LivebessV1:
```
Comment (DeusAnimo) [inline lunavl/sdk/estimator_collections.py]:
```suggestion
        elif estimator == FaceEstimator.LivenessV1:
```

File: lunavl/sdk/estimators/face_estimators/orientation_mode.py (hunk 1 above)
Comment (gluhar2006) [reply lunavl/sdk/estimators/face_estimators/orientation_mode.py]: updated

File: lunavl/sdk/estimators/face_estimators/orientation_mode.py (hunk 2 above)
Comment (gluhar2006) [reply lunavl/sdk/estimators/face_estimators/orientation_mode.py]: added

File: lunavl/sdk/estimators/face_estimators/orientation_mode.py (hunk 3 above)
Comment (gluhar2006) [reply lunavl/sdk/estimators/face_estimators/orientation_mode.py]: updated | image estimator

File: lunavl/sdk/estimators/face_estimators/orientation_mode.py (hunk 4 above)
Comment (gluhar2006) [reply lunavl/sdk/estimators/face_estimators/orientation_mode.py]: done

File: lunavl/sdk/faceengine/setting_provider.py (hunk 5 above)
Comment (gluhar2006) [reply lunavl/sdk/faceengine/setting_provider.py]: checked

File: tests/test_estimate_descriptor.py (hunk 6 above)
Comment (gluhar2006) [reply tests/test_estimate_descriptor.py]: added

File: lunavl/sdk/estimator_collections.py (hunk 7 above)
Comment (gluhar2006) [reply lunavl/sdk/estimator_collections.py]: fixed

File: lunavl/sdk/estimator_collections.py (hunk 8 above)
Comment (gluhar2006) [reply lunavl/sdk/estimator_collections.py]: fixed

File: docs/_sources/sdk/face_estimations.rst.txt [hunk 9]
```diff
@@ -209,3 +235,9 @@ Classes and methods
 
 .. automodule:: lunavl.sdk.estimators.face_estimators.glasses
     :members:
+
+.. automodule:: lunavl.sdk.estimators.face_estimators.livenessv1
+    :members:
+
+.. automodule:: lunavl.sdk.estimators.face_estimators.orientation_mode
```
Comment (DeusAnimo) [inline docs/_sources/sdk/face_estimations.rst.txt]: incorrect module import

File: lunavl/sdk/detectors/humandetector.py [hunk 10]
```diff
@@ -134,18 +141,31 @@ def detectOne(
...
 
         if detectArea is None:
             forDetection = ImageForDetection(image=image, detectArea=image.rect)
         else:
             forDetection = ImageForDetection(image=image, detectArea=detectArea)
         imgs, detectAreas = getArgsForCoreDetectorForImages([forDetection])
-        error, detectRes = self._detector.detect(
-            [imgs[0]], [detectAreas[0]], 1, self._getDetectionType(detectLandmarks)
-        )
+        error, detectRes = self._detector.detect(imgs, detectAreas, 1, detectionType)
```
Comment (Zeskbest) [inline lunavl/sdk/detectors/humandetector.py]: `1,` - what kinda magic variable is it?

File: lunavl/sdk/estimator_collections.py [hunk 11]
```diff
@@ -14,6 +14,8 @@
 from .estimators.face_estimators.mask import MaskEstimator
 from .estimators.face_estimators.glasses import GlassesEstimator
 from .estimators.face_estimators.facewarper import FaceWarper
+from .estimators.face_estimators.livenessv1 import LivenessV1Estimator
+from lunavl.sdk.estimators.image_estimators.orientation_mode import OrientationModeEstimator
```
Comment (Zeskbest) [inline lunavl/sdk/estimator_collections.py]:
```suggestion
from .estimators.image_estimators.orientation_mode import OrientationModeEstimator
```

File: lunavl/sdk/estimators/face_estimators/livenessv1.py [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+    def estimate(self, faceDetection: FaceDetection, headPose: HeadPose) -> LivenessV1:
+        """
+        Estimate liveness by detection with head pose validation
+        """
+        ...
+
+    @overload
+    def estimate(self, faceDetection: FaceDetection, yaw: float, pitch: float, roll) -> LivenessV1:
+        """
+        Estimate liveness by detection with head pose validation
```
Comment (Zeskbest) [inline lunavl/sdk/estimators/face_estimators/livenessv1.py]:
```suggestion
        Estimate liveness by detection with angles validation
```

File: lunavl/sdk/faceengine/setting_provider.py [hunk 13]
```diff
@@ -209,72 +209,45 @@ def setValue(self, name: str, value: Any) -> None:
...
-            name: setting name
-
-        Returns:
-            a float
-        """
-        return self.getValue(name).asFloat()
+        value = self._coreSettingProvider.getValue(self.__class__.sectionName, name)
+        if value is []:
+            return None
+        return value[0]
```
Comment (Zeskbest) [inline lunavl/sdk/faceengine/setting_provider.py]:
```suggestion
        value = self._coreSettingProvider.getValue(self.__class__.sectionName, name)
        if not value:
            return None
        return value[0]
```
image

File: lunavl/sdk/detectors/base.py [hunk 14]
```diff
@@ -120,4 +117,51 @@ def getArgsForCoreDetectorForImages(
...
+        coreImages.append(image.image.coreImage)
+        detectAreas.append([Detection(bbox.coreRect, 1.0) for bbox in image.bBoxes])
+
+    return coreImages, detectAreas
+
+
+def collectAndRaiseError(
+    error: FSDKErrorResult,
+    coreImages: List[CoreImage],
+    detectAreas: List[CoreImage],
```
Comment (DeusAnimo) [inline lunavl/sdk/detectors/base.py]: wrong typing

**Review by Zeskbest — CHANGES_REQUESTED** [review]

File: docs/_sources/sdk/face_estimations.rst.txt (hunk 9 above)
Comment (gluhar2006) [reply docs/_sources/sdk/face_estimations.rst.txt]: fixed

File: lunavl/sdk/detectors/base.py (hunk 14 above)
Comment (gluhar2006) [reply lunavl/sdk/detectors/base.py]: fixed

File: lunavl/sdk/estimator_collections.py (hunk 11 above)
Comment (gluhar2006) [reply lunavl/sdk/estimator_collections.py]: fixed

File: lunavl/sdk/detectors/humandetector.py (hunk 10 above)
Comment (gluhar2006) [reply lunavl/sdk/detectors/humandetector.py]: upd: limit

File: lunavl/sdk/estimators/face_estimators/livenessv1.py (hunk 12 above)
Comment (gluhar2006) [reply lunavl/sdk/estimators/face_estimators/livenessv1.py]: fixed

File: lunavl/sdk/faceengine/setting_provider.py (hunk 13 above)
Comment (gluhar2006) [reply lunavl/sdk/faceengine/setting_provider.py]: fixed

File: lunavl/sdk/estimators/face_estimators/mouth_state.py [hunk 15]
```diff
@@ -46,7 +46,7 @@ def mouth(self) -> float:
         Returns:
             value in range [0, 1]
         """
-        return self._coreEstimation.mouth
+        return self._coreEstimation.opened
```
Comment (matemax) [inline lunavl/sdk/estimators/face_estimators/mouth_state.py:49]:
```
    @property
    def mouth(self) -> float: 
    ```
    i think that function must be renamed

File: lunavl/sdk/estimators/image_estimators/orientation_mode.py [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+        """
+        Init.
+
+        Args:
+            coreOrientationModeEstimator: core orientation mode estimator
+        """
+        super().__init__(coreOrientationModeEstimator)
+
+    @CoreExceptionWrap(LunaVLError.EstimationOrientationModeError)
+    def estimate(self, image: Union[VLImage, FaceWarp, FaceWarpedImage]) -> OrientationType:
```
Comment (matemax) [inline lunavl/sdk/estimators/image_estimators/orientation_mode.py:64]: create OrientationType in lunavl with docs and cast OrientationType to new type

File: lunavl/sdk/image_utils/image.py [hunk 17]
```diff
@@ -171,16 +188,39 @@ def __init__(
...
+                ndarray=array, inputColorFormat=inputColorFormat, colorFormat=colorFormat or ColorFormat.R8G8B8
             )
         else:
             raise TypeError(f"Bad image type: {type(body)}")
 
         self.source = body
         self.filename = filename
 
+    @classmethod
+    def getRotated(cls, image: "VLImage", angle: ImageAngle):
```
Comment (matemax) [inline lunavl/sdk/image_utils/image.py]: this function is named `rotate` usually

File: lunavl/sdk/luna_faces.py [hunk 18]
```diff
@@ -243,6 +249,30 @@ def gaze(self) -> GazeDirection:
...
+
+        Returns:
+            livenessv1
+        """
+        if self._liveness is None:
+            self._liveness = self.estimatorCollection.livenessV1Estimator.estimate(self, self.headPose)
+        return self._liveness
+
+    @property
+    def orientationMode(self) -> OrientationType:
```
Comment (matemax) [inline lunavl/sdk/luna_faces.py]: wtf, rotation for detection???

**Comment by matemax** [comment]
Need test for function rotate

File: lunavl/sdk/estimators/face_estimators/mouth_state.py (hunk 15 above)
Comment (gluhar2006) [reply lunavl/sdk/estimators/face_estimators/mouth_state.py:49]: renamed

File: lunavl/sdk/estimators/image_estimators/orientation_mode.py (hunk 16 above)
Comment (gluhar2006) [reply lunavl/sdk/estimators/image_estimators/orientation_mode.py:64]: created

File: lunavl/sdk/image_utils/image.py (hunk 17 above)
Comment (gluhar2006) [reply lunavl/sdk/image_utils/image.py]: ok

File: lunavl/sdk/luna_faces.py (hunk 18 above)
Comment (gluhar2006) [reply lunavl/sdk/luna_faces.py]: removed

**Review by gluhar2006 — COMMENTED** [review]
> Need test for function rotate

added 4 simple tests

File: tests/test_rotation.py [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+        rotatedImage = VLImage.rotate(self.image, ImageAngle.ANGLE_270)
+        orientationMode = self.orientationModeEstimator.estimate(rotatedImage)
+        assert orientationMode == OrientationType.RIGHT
+
+    def test_rotate_upside_down(self):
+        """
+        Test image rotation - 180 angle
+        """
+        rotatedImage = VLImage.rotate(self.image, ImageAngle.ANGLE_180)
+        orientationMode = self.orientationModeEstimator.estimate(rotatedImage)
```
Comment (matemax) [inline tests/test_rotation.py]: compare rotate imageand real rotaded image and assert bytes. Use png image

File: tests/test_rotation.py [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+    def setup_class(cls):
+        super().setup_class()
+        cls.orientationModeEstimator = cls.faceEngine.createOrientationModeEstimator()
+        cls.image = VLImage.load(filename=ROTATED0)
+
+    def test_rotate_normal(self):
+        """
+        Test image rotation - 0 angle
+        """
+        rotatedImage = VLImage.rotate(self.image, ImageAngle.ANGLE_0)
```
Comment (matemax) [inline tests/test_rotation.py]: use subbtests

File: tests/test_orientation_mode.py [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+        ), f"{estimatedOrientation.__class__} is not {OrientationType}"
+        assert (
+            estimatedOrientation == expectedOrientation
+        ), f"Expected {expectedOrientation.value}, got {estimatedOrientation}"
+
+    def test_orientation_mode_normal(self):
+        """
+        Test orientation mode with normal rotation
+        """
+        orientationMode = TestOrientationMode.orientationModeEstimator.estimate(self.normalImage)
```
Comment (matemax) [inline tests/test_orientation_mode.py]: subtests?

**Comment by matemax** [comment]
Update version in pyproject toml (minor)

File: tests/test_rotation.py (hunk 19 above)
Comment (gluhar2006) [reply tests/test_rotation.py]: ok

File: tests/test_rotation.py (hunk 20 above)
Comment (gluhar2006) [reply tests/test_rotation.py]: ok

File: tests/test_orientation_mode.py (hunk 21 above)
Comment (gluhar2006) [reply tests/test_orientation_mode.py]: ok

**Review by Zeskbest — APPROVED** [review]

**Pull request merged**